rfc/halyard-lite.md

```diff
@@ -0,0 +1,475 @@
+This is a fairly representative example of a config parameter; it supports add/edit/delete/get/list
+on a list of Google Pubsub subscriptions, and required 2244 lines of code to
+implement (none of which is test code).
+
+The amount of code necessary to add simple config parameters is a burden both on
+contributors and on the maintainers of Halyard who have to review this code.  This
+has led to Halyard’s config often lagging behind that of the individual microservices.
+Ideally, adding a config parameter would be a reasonably simple first contribution
+for someone looking to get involved with the project, but that is currently very
+far from the case.
+
+## Timeline
+
+* Spinnaker 1.19 (March 2020)
+  * An alpha version of Halyard Lite is released along with documentation, and
+  is able to generate service configurations from the Halyard config.
+  * A simple Kustomize kustomization is released that can deploy Spinnaker to
+  Kubernetes in simple cases.
+* Spinnaker 1.20 (May 2020)
+  * Feedback from early users of Halyard Lite and the kustomization are
+   addressed, and the kustomization now supports more advanced use cases (ex:
+   high availability, custom BOM, etc.)
+  * The Armory operator is moved to depend on Halyard Lite
+* Spinnaker 1.21 (July 2020)
+  * Halyard Lite, as well as the kustomize install path are GA and the
+   recommended install path for new users of Spinnaker.
+  * Documentation (such as install documentation) referencing old Halyard
+  commands is updated to reference Halyard Lite.
+  * Based on adoption of Halyard Lite, a plan is formulated around how and when
+  to deprecate and cease support of Halyard. The details of this plan will be a
+  separate RFC.
+
+## Design
+
+### Removed and replaced features
+
+#### Commands to edit the Halyard config
+
+There are currently two ways to update the Halyard configuration:
+*   Edit the YAML directly
+*   Run `hal …` commands to update the config
+
+YAML is ubiquitous in the Kubernetes world, and Kubernetes-native tools generally
+assume that users are comfortable writing YAML configuration files. As such, the
+value in providing a CLI that hides YAML from users is questionable, and may even
+contribute to the perception that Spinnaker (or at least its installation) is not
+Kubernetes-native.  The CLI also encourages an imperative workflow whereby
+configuration is achieved via a series hal commands.  Furthermore, a significant
+amount of the complexity and verbosity of Halyard’s codebase comes from the commands
+to edit the Halyard configuration.
+
+Based on these considerations, the CLI commands to edit the configuration will be
+removed.  As the [hal command reference](https://www.spinnaker.io/reference/halyard/commands/)
+is the best existing documentation on the available config parameters for Spinnaker,
+it will be replaced by extensive documentation of a well-typed Halyard configuration
+so that users can be confident directly editing this config.
```

Comment:
Yes, this is an important part of deploying to Kubernetes.  Halyard Lite will stage the files just as Halyard did, but it will then be the responsibility of Helm/Kustomize to create config maps/secrets out of those files and deploy them to the cluster.

I think this actually opens up for better Kubernetes-native solutions going forward...such as just directly referencing a secret in the config file and decoupling the deployment of Spinnaker from the deployment of the tokens that it uses.

Reply:
Added managing files to the list of things that were complex pre-Halyard as this is a good point.